debezium-api/src/main/java/io/debezium/engine/reactive/ReactiveDebeziumEngine.java

```diff
@@ -0,0 +1,225 @@
+/*
+ * Copyright Debezium Authors.
+ *
+ * Licensed under the Apache Software License version 2.0, available at http://www.apache.org/licenses/LICENSE-2.0
+ */
+package io.debezium.engine.reactive;
+
+import io.debezium.DebeziumException;
+import io.debezium.common.annotation.Incubating;
+import io.debezium.engine.ChangeEvent;
+import io.debezium.engine.RecordChangeEvent;
+import io.debezium.engine.format.ChangeEventFormat;
+import io.debezium.engine.format.KeyValueChangeEventFormat;
+import io.debezium.engine.format.SerializationFormat;
+import io.debezium.engine.spi.OffsetCommitPolicy;
+import org.reactivestreams.Publisher;
+import org.slf4j.LoggerFactory;
+
+import java.io.Closeable;
+import java.time.Clock;
+import java.util.*;
+import java.util.function.Consumer;
+
+@Incubating
+public interface ReactiveDebeziumEngine<R> extends Closeable {
+    public static final String OFFSET_FLUSH_INTERVAL_MS_PROP = "offset.flush.interval.ms";
+
+    public interface CompletionCallback {
+        /**
+         * Handle the completion of the embedded connector engine.
+         *
+         * @param success {@code true} if the connector completed normally, or {@code false} if the connector produced an error
+         *            that prevented startup or premature termination.
+         * @param message the completion message; never null
+         * @param error the error, or null if there was no exception
+         */
+        void handle(boolean success, String message, Throwable error);
+    }
+
+    /**
+     * Callback function which informs users about the various stages a connector goes through during startup
+     */
+    public interface ConnectorCallback {
+
+        /**
+         * Called after a connector has been successfully started by the engine; i.e. {@link SourceConnector#start(Map)} has
+         * completed successfully
+         */
+        default void connectorStarted() {
+            // nothing by default
+        }
+
+        /**
+         * Called after a connector has been successfully stopped by the engine; i.e. {@link SourceConnector#stop()} has
+         * completed successfully
+         */
+        default void connectorStopped() {
+            // nothing by default
+        }
+
+        /**
+         * Called after a connector task has been successfully started by the engine; i.e. {@link SourceTask#start(Map)} has
+         * completed successfully
+         */
+        default void taskStarted() {
+            // nothing by default
+        }
+
+        /**
+         * Called after a connector task has been successfully stopped by the engine; i.e. {@link SourceTask#stop()} has
+         * completed successfully
+         */
+        default void taskStopped() {
+            // nothing by default
+        }
+    }
+
+    /**
+     * A builder to set up and create {@link ReactiveDebeziumEngine} instances.
+     */
+    public static interface Builder<R> {
```

Comment:
Could we have a single builder for both engine flavours and add a method like `reactive()` so it returns one or the other?